### ISSUE TITLE
Target paths fail to resolve relative to project

### ISSUE THREAD

**Issue by JacobLey** [nrwl/nx#19202]
### Current Behavior

`packages/my-package/package.json`:
```
{
    "name": "my-package",
    "devDependencies": {
        "my-plugin": "1.2.3"
    }
}
```

`packages/my-package/project.json`:
```
{
  "name": "my-package",
  "targets": {
    "my-target": {
      "executor": "my-plugin:do-stuff",
      "options": {}
    }
  }
}

```

Running command:
`nx run my-package:my-target`

Results in error:
```
Unable to resolve my-package:my-target.

unable to find tsconfig.base.json or tsconfig.json
```

(Side note, this error is horribly unhelpful)

### Expected Behavior

Plugin/target should execute as expected, because it successfully resolves the dependency relative to the project

### Steps to Reproduce

1. See current behavior

### Nx Report

```shell
Node   : 20.5.1
   OS     : darwin-x64
   pnpm   : 8.7.4
   
   nx (global)  : 16.7.2
   nx           : 16.8.1
   @nrwl/tao    : 16.7.2
   typescript   : 5.2.2
   ---------------------------------------
   The following packages should match the installed version of nx
     - @nrwl/tao@16.7.2
   
   To fix this, run `nx migrate nx@16.8.1`
```

### Package Manager Version

pnpm: 8.7.4

### Operating System

- [X] macOS
- [ ] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

https://github.com/nrwl/nx/blob/master/packages/nx/src/command-line/run/executor-utils.ts#L112

The error _really_ occurs here, as it does not ever attempt to resolve the dependency relative to the package itself.
It only tries the root (expecting global installations, a general antipattern for monorepos), or the cwd (inconsistent) or some nx-specific directories.

However the misleading error occurs because this miss doesn't _immediately_ trigger an error, and it instead falls back to  https://github.com/nrwl/nx/blob/master/packages/nx/src/utils/nx-plugin.ts#L460

**Comment by AgentEnder** [comment]
Nx plugins need to be installed in the workspace root, thats a current limitation but not one that I see changing any time soon. I'm going to close this out as it is currently working as intended, although I do concede that the error message isn't super helpful.

**Comment by JacobLey** [comment]
Closing the issue while acknowledging it exists seems like the wrong approach in my opinion.

If it helps to rephrase this as a feature request, I am open to that.

Is there a justification for not supporting plugins as direct dependencies of packages? If it simply isn't supported yet, I have linked to the relevant code above and would be open to attempting a PR.

It causes issues when developing local plugins (as plugins needs to be treated as a regular project dependency) and stated above hoisting to the workspace root is generally an anti-pattern in monorepos, as it leads to leaky dependencies.

**Comment by AgentEnder** [comment]
> Closing the issue while acknowledging it exists seems like the wrong approach in my opinion.

I see where you are coming from on this, but if we acknowledge that it is a limitation and that we don't plan to fix it currently then it doesn't make sense to keep it open either. To me an open issue is something that is one of the below:
- A potential bug (not yet verified)
- A bug that we are working on fixing
- A feature request thats not been reviewed
- A feature request that we are either working on, would like to work on, or would accept a pull request for

If an issue is determined to be any of the below, it makes more sense to close it to limit confusion:
- A feature request that we don't wish to add to Nx for any reason (maintenance, direction, etc)
- A known "bug" that is more of a limitation. In these cases, we are talking about things where its something that doesn't work that someone may have expected to work, but nothing is actually operating badly.

These are just my takes, they aren't necessarily correct and aren't necessarily shared by the whole team. Nonetheless, its how I typically handle issue triage.

> Is there a justification for not supporting plugins as direct dependencies of packages? If it simply isn't supported yet, I have linked to the relevant code above and would be open to attempting a PR.

Yeah, so one of the big things is that when loading executors we support "batch mode". When running Nx with batch mode, which is soon to be turned on by default, some executors will load once per run instead of once per task. In batch mode, targets that run via that same executor are ran in one execution process and you see significant speed increases. Currently this is supported for `tsc` and `jest`, but I'd wager more will come in the future.

If we open up support for loading executors from within that specific project's node modules, batch mode gets a bit hairy. `@nx/js:tsc` would not be able to support loading in from the project root. 

We could theoretically support this for **_non-batch_** executors, but then its also confusing why it works for some executors but not others, why it wouldn't work for generators (which are not project specific), and why graph plugins also wouldn't support it (they aren't project specific either).

It makes more sense and is more intuitive if we simply establish that as a rule, plugins are installed to the workspace rather than a project.

> It causes issues when developing local plugins (as plugins needs to be treated as a regular project dependency) and stated above hoisting to the workspace root is generally an anti-pattern in monorepos, as it leads to leaky dependencies.

I'm not sure what issue you are hitting. If something with local plugin execution isn't working, you should file an issue for that in particular. If the plugin is developed in the same workspace its used in, it shouldn't have to be installed as we invoke it via ts-node or swc-node (if available).

**Comment by JacobLey** [comment]
> if we acknowledge that it is a limitation and that we don't plan to fix it currently then it doesn't make sense to keep it open either

Fair enough.

> In batch mode, targets that run via that same executor are ran in one execution process

I am vaguely familiar with batch mode, but not up to date with the internals. However if I had to guess, I expect it de-dupes executors based on some "key". Most likely the literal string declared in the target.

I understand the initial concern if that if each dependency is resolved separately, which resolved plugin runs the batch?

I would suggest instead differing the dedupe-ing of targets until _after_ resolution. Due to the way package managers like pnpm expose dependencies, so long as separate packages share similar dependencies, they will resolve to the exact same instance of the plugin!

e.g.
```ts
// Next targets to run in the dependency "queue"
const projects = [
    { name: 'lib-A', executor: '@nx/js:tsc' },
    { name: 'lib-B', executor: '@nx/noop' },
    { name: 'lib-C', executor: '@nx/js:tsc' },
];

// "current way" (heavy pseudocode)
const batchedProjects = new Map<string, string>();
for (const { name , executor } of projects) {
    const projectBatch = batchedProjects.get(executor) ?? [];
    projectBatch.push(name);
    batchedProjects.set(executor, projectBatch);
}
for (const [executor, projects] of batchedProjects.entries()) {
   // This resolves _after_ grouping, so it doesn't make sense to resolve per-project.
    const resolved = resolveExecutor(executor);
    resolved.batchMode(projects);
}

// Alternative approach
const batchedProjects = new Map<ResolvedExecutor, string>();
for (const { name , executor } of projects) {
    const resolvedExecutor = resolveExecutorFromProject(executor);
    const projectBatch = batchedProjects.get(resolvedExecutor) ?? [];
    projectBatch.push(name);
    batchedProjects.set(resolvedExecutor projectBatch);
}
for (const [resolved, projects] of batchedProjects.entries()) {
    // already resolved, so doesn't matter if it came from root or project!
    resolved.batchMode(projects);
}
```

This may actually be incredibly valuable for the case where a project may need to peg their dependency on a plugin to an older version, but don't want to hold back the other packages in the workspace.

We are getting more into feature request area instead of issue, but again would be open to attempting a PR if this is reasonable.

> but then its also confusing why it works for some executors but not others

FWIW I would argue the current behavior of resolving from `process.cwd()` falls under this category.

If my cwd is a project directory, plugins _not_ installed at workspace root may suddenly magically work. Then suddenly break again when I change my directory. This is non-obvious to the end user, because the nx cli _should_ work the same from any directory in the workspace. 

Even if resolving from projects were supported, I would vote to remove the cwd resolution.

> If something with local plugin execution isn't working

I've mostly been using the approach of treating a plugin as a normal dependency (so it gets built + tested before ever executing). I haven't tried relying on `ts-node` (was not aware of this support).

I have had good enough success just declaring the local dependency in workspace root, and _also_ adding it to each project's `devDependencies` (or `implicitDependencies`, but I prefer a more explicit approach).
